core/src/main/scala/org/http4s/client/jdkhttpclient/WebSocketClient.scala

```diff
@@ -0,0 +1,167 @@
+  def sendMany[G[_]: Traverse, A <: WSDataFrame](wsfs: G[A]): F[Unit]
+  final def sendPipe: Pipe[F, WSDataFrame, Unit] = _.chunks.evalMap(sendMany(_))
+
+  /** Send a Ping frame. */
+  def sendPing(data: ByteVector = ByteVector.empty): F[Unit]
+
+  /** Send a Close frame. The sending side of this connection will be closed. */
+  def sendClose(reason: String = ""): F[Unit]
+
+  /** A stream of the incoming websocket frames. */
+  def receiveStream: Stream[F, WSDataFrame]
+
+  /** The negotiated subprotocol, if any. */
+  def subprocotol: Option[String]
+
+  /** The close frame, if available. */
+  def closeFrame: TryableDeferred[F, WSFrame.Close]
+
+}
+
+trait WebSocketClient[F[_]] {
+
+  /** Establish a websocket connection. It will be closed automatically if necessary. */
+  def connect(request: WSRequest): Resource[F, WSConnection[F]]
+
+  /** Establish a "high level" websocket connection. You only get to handle Text and Binary frames.
+    * Pongs will be replied automatically. Received frames are grouped by the `last` attribute. The
+    * connection will be closed automatically.
+    */
+  def connectHighLevel(request: WSRequest): Resource[F, WSConnectionHighLevel[F]]
+}
+
+object WebSocketClient {
+
+  def defaultImpl[F[_]](
+      respondToPings: Boolean
+  )(f: WSRequest => Resource[F, WSConnection[F]])(implicit F: Concurrent[F]): WebSocketClient[F] =
+    new WebSocketClient[F] {
+      override def connect(request: WSRequest) = f(request)
+      override def connectHighLevel(request: WSRequest) =
+        for {
+          (recvCloseFrame, outputOpen) <- Resource.liftF(
+            Deferred.tryable[F, WSFrame.Close].product(Ref[F].of(false))
+          )
+          conn <- f(request)
+        } yield
+          new WSConnectionHighLevel[F] {
+            override def send(wsf: WSDataFrame) = conn.send(wsf)
+            override def sendMany[G[_]: Traverse, A <: WSDataFrame](wsfs: G[A]): F[Unit] =
+              conn.sendMany(wsfs)
+            override def sendPing(data: ByteVector) = conn.send(WSFrame.Ping(data))
+            override def sendClose(reason: String) =
+              conn.send(WSFrame.Close(1000, reason)) *> outputOpen.set(false)
+            override def receiveStream: Stream[F, WSDataFrame] =
+              conn.receiveStream
+                .evalTap {
+                  case WSFrame.Ping(data) if respondToPings => conn.send(WSFrame.Pong(data))
+                  case wsf: WSFrame.Close =>
+                    recvCloseFrame.complete(wsf) *> outputOpen.get.flatMap(conn.send(wsf).whenA(_))
+                  case _ => F.unit
+                }
+                .collect { case wsdf: WSDataFrame => wsdf }
+                .through(groupFrames[F])
+            override def subprocotol: Option[String] = conn.subprotocol
+            override def closeFrame: TryableDeferred[F, WSFrame.Close] = recvCloseFrame
+          }
+    }
+
+  private[jdkhttpclient] def groupFrames[F[_]]: Pipe[F, WSDataFrame, WSDataFrame] = {
```

Comment:
I could not come up with a straightforward way to write this with the existing combinators...